rectools/visuals/metrics_app.py

```diff
@@ -85,125 +104,275 @@ def construct(
+
+    @staticmethod
+    def _validate_models_metadata(models_metadata: pd.DataFrame) -> None:
+        if Columns.Model not in models_metadata.columns:
+            raise KeyError("Missing `Model` column in `models_metadata` DataFrame")
+        if models_metadata[Columns.Model].isnull().any():
+            raise ValueError("Found NaN values in `Model` column")
+        if models_metadata[Columns.Model].nunique() != len(models_metadata):
+            raise ValueError("`Model` values of `models_metadata` should be unique`")
 
     @lru_cache
-    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:  # pragma: no cover
-        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+    def _make_chart_data_fold(self, fold_number: int) -> pd.DataFrame:
+        return self.data[self.data[Columns.Split] == fold_number].reset_index(drop=True)
 
     @lru_cache
     def _make_chart_data_avg(self) -> pd.DataFrame:
-        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+        metric_data_columns = [Columns.Model] + self.metric_names
+        meta_data_columns = [Columns.Model] + self.meta_names
+        # separate metric data because meta data could contain strings
+        metrics_data = self.data[metric_data_columns].groupby([Columns.Model], sort=False).mean().reset_index()
+        meta_data = self.data[meta_data_columns].drop_duplicates()
+        return metrics_data.merge(meta_data, on=Columns.Model, how="left").reset_index(drop=True)
 
-    def _update_chart(
+    @staticmethod
+    @lru_cache
+    def _split_to_meta_and_model(raw_string: str, sep: str = META_MODEL_SEP) -> tp.Tuple[str, str]:
+        splitted_row = raw_string.split(sep, 1)
+        if len(splitted_row) > 1:
+            meta_value, model_name = splitted_row
+            return meta_value, model_name
+        return "", raw_string
+
+    def _create_chart_figure(
+        self,
+        data: pd.DataFrame,
+        x_col: str,
+        y_col: str,
+        color_col: str,
+        legend_title: str,
+    ) -> go.Figure:  # pragma: no cover
+        scatter_kwargs = {
+            "width": WIDGET_WIDTH,
+            "height": WIDGET_HEIGHT,
+        }
+        scatter_kwargs.update(self.scatter_kwargs)
+
+        data = data.sort_values(by=color_col, ascending=True)
+        data[color_col] = data[color_col].astype(str)  # to treat colors values as categorical
+
+        fig = px.scatter(
+            data,
+            x=x_col,
+            y=y_col,
+            color=color_col,
+            symbol=Columns.Model,
+            **scatter_kwargs,
+        )
+        # Set custom legend for meta info coloring
+        if color_col != Columns.Model:
+            for trace, meta_value, model_name in zip(fig.data, data[color_col], data[Columns.Model]):
+                trace.name = f"{meta_value}{META_MODEL_SEP}{model_name}"
```

Comment:
Just out of curiosity: didn't `labels` argument help?

Reply:
And let's maybe swap model name and meta value

Reply:
Please don't swap. Meta value will not be readable if we swap

Reply:
discussed, agreed not to swap